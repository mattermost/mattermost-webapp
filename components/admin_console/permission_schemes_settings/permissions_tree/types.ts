// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export type Permission = {
    id: string;
    combined?: boolean;
    permissions: string[];
}
export type Group = {
    id: string;
    permissions: Array<Permission | string>;
}

export type AdditionalValues = {
    edit_post: {
        editTimeLimitButton: JSX.Element;
    };
}
