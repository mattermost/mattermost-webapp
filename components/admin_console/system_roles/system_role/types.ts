// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export type ReadAccess = 'read';
export const readAccess: ReadAccess = 'read';
export type WriteAccess = 'write';
export const writeAccess: WriteAccess = 'write';
export type NoAccess = false;
export const noAccess: NoAccess = false;
export type MixedAccess = 'mixed';
export const mixedAccess: MixedAccess = 'mixed';

export type PermissionAccess = ReadAccess | WriteAccess | NoAccess;
export type PermissionsToUpdate = Record<string, ReadAccess | WriteAccess | NoAccess>;
export type PermissionToUpdate = {
    name: string;
    value: PermissionAccess;
};

export type SystemSection = {
    name: string;
    hasDescription?: boolean;
    subsections?: SystemSection[];
    disabled?: boolean;
}
