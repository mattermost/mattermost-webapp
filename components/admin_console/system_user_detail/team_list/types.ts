// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Team, TeamMembership} from '@mattermost/types/teams';

export type TeamWithMembership = Team & TeamMembership;
