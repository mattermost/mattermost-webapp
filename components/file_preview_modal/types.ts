// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {FileInfo} from '@mattermost/types/files';

export type LinkInfo = {
    has_preview_image: boolean;
    link: string;
    extension: string;
    name: string;
}

export function isFileInfo(info: FileInfo | LinkInfo): info is FileInfo {
    return Boolean((info as FileInfo).id);
}
