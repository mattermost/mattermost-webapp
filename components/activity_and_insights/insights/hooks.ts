// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {useSelector} from 'react-redux';

import {getCloudSubscription, getSubscriptionProduct} from 'mattermost-redux/selectors/entities/cloud';
import {getConfig, getLicense} from 'mattermost-redux/selectors/entities/general';

import {isCloudLicense} from 'utils/license_utils';
import {CloudProducts, InsightsScopes} from 'utils/constants';

import {setGlobalItem} from 'actions/storage';

import {useGlobalState} from 'stores/hooks';

/**
 * Returns some checks for free trial users or starter licenses
 */
export function useLicenseChecks(): {isStarterFree: boolean; isFreeTrial: boolean; isEnterpriseReady: boolean} {
    const subscription = useSelector(getCloudSubscription);
    const license = useSelector(getLicense);
    const subscriptionProduct = useSelector(getSubscriptionProduct);
    const config = useSelector(getConfig);

    const isCloud = isCloudLicense(license);
    const isCloudStarterFree = isCloud && subscriptionProduct?.sku === CloudProducts.STARTER;
    const isCloudFreeTrial = isCloud && subscription?.is_free_trial === 'true';

    const isEnterpriseReady = config.BuildEnterpriseReady === 'true';
    const isSelfHostedStarter = isEnterpriseReady && (license.IsLicensed === 'false');
    const isSelfHostedFreeTrial = license.IsTrial === 'true';

    const isStarterFree = isCloudStarterFree || isSelfHostedStarter;
    const isFreeTrial = isCloudFreeTrial || isSelfHostedFreeTrial;

    return {
        isStarterFree,
        isFreeTrial,
        isEnterpriseReady,
    };
}

export function useGetFilterType(): [string, (value: string) => ReturnType<typeof setGlobalItem>] {
    const {isStarterFree, isEnterpriseReady} = useLicenseChecks();
    const [filterType, setFilterType] = useGlobalState(InsightsScopes.TEAM, 'insightsScope');

    let returnedFilterType = filterType;

    // Force MY Insights for free trial users
    if (isStarterFree || !isEnterpriseReady) {
        returnedFilterType = InsightsScopes.MY;
    }

    return [
        returnedFilterType,
        setFilterType,
    ];
}
