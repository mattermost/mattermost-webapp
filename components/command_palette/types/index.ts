// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export enum CommandPaletteEntities {
    Channel = 'channel',
    Playbooks = 'playbooks',
    Boards = 'boards',
    Messages = 'messages',
    Files = 'files',
    Runs = 'runs',
    GoTo = 'goTo',
}

export enum ChannelType {
    DM_CHANNEL= 'D',
    GM_CHANNEL= 'G',
    OPEN_CHANNEL= 'O',
    PRIVATE_CHANNEL= 'P',
}

export enum BoardsType {
    OPEN_CHANNEL= 'O',
    PRIVATE_CHANNEL= 'P',
}

export enum PlaybooksType {
    OPEN_CHANNEL= 'O',
    PRIVATE_CHANNEL= 'P',
}

