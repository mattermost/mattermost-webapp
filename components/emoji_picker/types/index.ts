// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {EmojiCategory, Emoji} from 'mattermost-redux/types/emojis';

export type Category = {
    className: string;
    emojiIds?: string[];
    id: string;
    message: string;
    name: EmojiCategory;
    offset: number;
};

export type Categories = Record<EmojiCategory, Category>;

export type CategoryOrEmojiRow = {
    type: 'categoryHeaderRow' | 'emojisRow';
    row: EmojiCategory | Emoji[];
}
