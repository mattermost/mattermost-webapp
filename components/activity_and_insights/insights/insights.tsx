// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {memo, useEffect, useCallback} from 'react';
import {useDispatch, useSelector} from 'react-redux';

import {trackEvent} from 'actions/telemetry_actions';
import {suppressRHS, unsuppressRHS} from 'actions/views/rhs';
import LocalStorageStore from 'stores/local_storage_store';
import {useGlobalState} from 'stores/hooks';

import {InsightsScopes, PreviousViewedTypes, suitePluginIds} from 'utils/constants';
import {useProducts} from 'utils/products';

import {getCurrentUserId} from 'mattermost-redux/selectors/entities/users';
import {getCurrentTeamId} from 'mattermost-redux/selectors/entities/teams';

import {selectLhsItem} from 'actions/views/lhs';
import {GlobalState} from 'types/store';
import {LhsItemType, LhsPage} from 'types/store/lhs';

import {CardSizes, InsightsWidgetTypes, TimeFrame, TimeFrames} from '@mattermost/types/insights';

import InsightsHeader from './insights_header/insights_header';
import TopChannels from './top_channels/top_channels';
import TopReactions from './top_reactions/top_reactions';
import TopThreads from './top_threads/top_threads';
import TopBoards from './top_boards/top_boards';
import LeastActiveChannels from './least_active_channels/least_active_channels';
import TopPlaybooks from './top_playbooks/top_playbooks';
import TopDMsAndNewMembers from './top_dms_and_new_members/top_dms_and_new_members';

import {useGetFilterType} from './hooks';

import './../activity_and_insights.scss';

type SelectOption = {
    value: string;
    label: string;
}

const Insights = () => {
    const dispatch = useDispatch();

    // check if either of focalboard plugin or boards product is enabled
    const focalboardPluginEnabled = useSelector((state: GlobalState) => state.plugins.plugins?.focalboard);
    let focalboardProductEnabled = false;
    const products = useProducts();
    if (products) {
        focalboardProductEnabled = products.some((product) => product.pluginId === suitePluginIds.focalboard || product.pluginId === suitePluginIds.boards);
    }
    const focalboardEnabled = focalboardPluginEnabled || focalboardProductEnabled;

    const playbooksEnabled = useSelector((state: GlobalState) => state.plugins.plugins?.playbooks);
    const currentUserId = useSelector(getCurrentUserId);
    const currentTeamId = useSelector(getCurrentTeamId);

    const [filterType, setFilterType] = useGetFilterType();
    const [timeFrame, setTimeFrame] = useGlobalState(TimeFrames.INSIGHTS_7_DAYS as string, 'insightsTimeFrame');

    const setFilterTypeTeam = useCallback(() => {
        trackEvent('insights', 'change_scope_to_team_insights');
        setFilterType(InsightsScopes.TEAM);
    }, []);

    const setFilterTypeMy = useCallback(() => {
        trackEvent('insights', 'change_scope_to_my_insights');
        setFilterType(InsightsScopes.MY);
    }, []);

    const setTimeFrameValue = useCallback((value: SelectOption) => {
        setTimeFrame(value.value);
    }, []);

    useEffect(() => {
        dispatch(selectLhsItem(LhsItemType.Page, LhsPage.Insights));
        dispatch(suppressRHS);
        const penultimateType = LocalStorageStore.getPreviousViewedType(currentUserId, currentTeamId);

        if (penultimateType !== PreviousViewedTypes.INSIGHTS) {
            LocalStorageStore.setPenultimateViewedType(currentUserId, currentTeamId, penultimateType);
            LocalStorageStore.setPreviousViewedType(currentUserId, currentTeamId, PreviousViewedTypes.INSIGHTS);
        }

        return () => {
            dispatch(unsuppressRHS);
        };
    }, []);

    return (
        <>
            <InsightsHeader
                filterType={filterType}
                setFilterTypeTeam={setFilterTypeTeam}
                setFilterTypeMy={setFilterTypeMy}
                timeFrame={timeFrame}
                setTimeFrame={setTimeFrameValue}
            />
            <div className='insights-body'>
                <TopChannels
                    size={CardSizes.large}
                    filterType={filterType}
                    widgetType={InsightsWidgetTypes.TOP_CHANNELS}
                    class={'top-channels-card'}
                    timeFrame={timeFrame as TimeFrame}
                />
                <div className='card-row'>
                    <TopThreads
                        size={focalboardEnabled ? CardSizes.small : CardSizes.medium}
                        filterType={filterType}
                        widgetType={InsightsWidgetTypes.TOP_THREADS}
                        class={'top-threads-card'}
                        timeFrame={timeFrame as TimeFrame}
                    />
                    {
                        focalboardEnabled &&
                        <TopBoards
                            size={CardSizes.small}
                            filterType={filterType}
                            widgetType={InsightsWidgetTypes.TOP_BOARDS}
                            class={'top-boards-card'}
                            timeFrame={timeFrame as TimeFrame}
                        />
                    }
                    <TopReactions
                        size={focalboardEnabled ? CardSizes.small : CardSizes.medium}
                        filterType={filterType}
                        widgetType={InsightsWidgetTypes.TOP_REACTIONS}
                        class={'top-reactions-card'}
                        timeFrame={timeFrame as TimeFrame}
                    />
                </div>
                <TopDMsAndNewMembers
                    size={CardSizes.large}
                    filterType={filterType}
                    widgetType={filterType === InsightsScopes.MY ? InsightsWidgetTypes.TOP_DMS : InsightsWidgetTypes.NEW_TEAM_MEMBERS}
                    class={'top-dms-card'}
                    timeFrame={timeFrame as TimeFrame}
                />
                <div className='card-row'>
                    <LeastActiveChannels
                        size={CardSizes.medium}
                        filterType={filterType}
                        widgetType={InsightsWidgetTypes.LEAST_ACTIVE_CHANNELS}
                        class={'least-active-channels-card'}
                        timeFrame={timeFrame as TimeFrame}
                    />
                    {
                        playbooksEnabled &&
                        <TopPlaybooks
                            size={CardSizes.medium}
                            filterType={filterType}
                            widgetType={InsightsWidgetTypes.TOP_PLAYBOOKS}
                            class={'top-playbooks-card'}
                            timeFrame={timeFrame as TimeFrame}
                        />
                    }
                </div>

            </div>
        </>
    );
};

export default memo(Insights);
