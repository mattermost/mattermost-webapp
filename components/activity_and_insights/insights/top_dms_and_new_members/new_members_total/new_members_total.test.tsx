// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {shallow} from 'enzyme';

import {TimeFrames} from '@mattermost/types/insights';

import NewMembersTotal from './new_members_total';

describe('components/activity_and_insights/insights/top_dms_and_new_members/new_members_total', () => {
    const props = {
        total: 2,
        timeFrame: TimeFrames.INSIGHTS_7_DAYS,
        openInsightsModal: jest.fn(),
    };

    test('should match snapshot with team', () => {
        const wrapper = shallow(
            <NewMembersTotal
                {...props}
            />,
        );
        expect(wrapper).toMatchSnapshot();
    });
});
