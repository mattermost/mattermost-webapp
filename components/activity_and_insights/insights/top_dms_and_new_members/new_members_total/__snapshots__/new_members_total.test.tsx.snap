// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/activity_and_insights/insights/top_dms_and_new_members/new_members_total should match snapshot with team 1`] = `
<div
  className="new-members-item new-members-info"
>
  <span
    className="total-count"
  >
    2
  </span>
  <div
    className="members-info"
  >
    <span
      className="time-range-info"
    >
      Joined the team in the last 7 days
    </span>
    <button
      className="see-all-button"
      onClick={[Function]}
    >
      See all
      <i
        className="icon icon-chevron-right"
      />
    </button>
  </div>
</div>
`;
